src/cache/cache_las.c

```diff
@@ -765,46 +698,123 @@ __wt_las_insert_updates(WT_CURSOR *cursor, WT_BTREE *btree, WT_PAGE *page, WT_MU
 
-            WT_ERR(__las_insert_record(session, cursor, btree_id, key, upd, &stop_ts, &stop_txnid));
-            ++insert_cnt;
+                prev_full_value->data = prev_upd->data;
+                prev_full_value->size = prev_upd->size;
+            }
 
             /*
-             * If we encounter subsequent updates with the same timestamp, skip them to avoid
-             * clobbering the entry that we just wrote to lookaside. Only the last thing in the
-             * update list for that timestamp and transaction id matters anyway.
+             * Skip the updates have the same start timestamp and transaction id
+             *
+             * The update older than onpage_upd can be squashed away. Insert a full update anyway to
+             * simplify the code. It will take some extra space but such case should be rare.
              */
-            while (upd->next != NULL && upd->start_ts == upd->next->start_ts &&
-              upd->txnid == upd->next->txnid)
-                upd = upd->next;
-        } while ((upd = upd->next) != NULL);
+            if (upd->start_ts != prev_upd->start_ts || upd->txnid != prev_upd->txnid ||
+              modifies.size == 0) {
+                /*
+                 * Calculate reverse delta. Insert full update for the newest historical record even
+                 * it's a MODIFY.
+                 *
+                 * It is not correct to check prev_upd == list->onpage_upd as we may have aborted
+                 * updates in the middle.
+                 */
+                if (upd->type == WT_UPDATE_MODIFY && modifies.size > 0) {
+                    nentries = MAX_REVERSE_MODIFY_NUM;
+                    if (__wt_calc_modify((WT_SESSION *)session, prev_full_value, full_value,
+                          prev_full_value->size / 10, entries, &nentries) == 0) {
+                        WT_ERR(__wt_modify_pack(cursor, &modify_value, entries, nentries));
+                        WT_ASSERT(session, modify_value != NULL);
+                        if (__las_insert_record(session, cursor, btree_id, key, upd,
+                              WT_UPDATE_MODIFY, modify_value, stop_ts_pair) != 0) {
+                            __wt_scr_free(session, &modify_value);
+                            goto err;
+                        } else
+                            __wt_scr_free(session, &modify_value);
+                    } else
+                        WT_ERR(__las_insert_record(session, cursor, btree_id, key, upd,
+                          WT_UPDATE_STANDARD, full_value, stop_ts_pair));
+                } else
+                    WT_ERR(__las_insert_record(session, cursor, btree_id, key, upd,
+                      WT_UPDATE_STANDARD, full_value, stop_ts_pair));
+
+                ++insert_cnt;
+            } else
+                squashed = true;
+        }
+
+        if (squashed)
+            WT_STAT_CONN_INCR(session, cache_lookaside_write_squash);
```

Comment:
I don't see any point in setting/testing the squashed variable, just hit the statistic where it's set to true.

Reply:
Do you suggest removing the statistic as well?

Reply:
I don't have an opinion on removing the statistic, I didn't think about whether it's useful or not.

I was only noting there was no reason to set/test the local variable, it was just as easy to update the statistic in the place we set the local variable.

Reply:
Ah, yeah. That's right. Will fix that.

Reply:
I suddenly realised removing the squashed local variable changed the behaviour. Now if we skip two updates, we count two squashes. Before with the squashed local variable, we count one. This is not that important, but for the sake of consistency, as we have another statistic which counts the number of squashed reads in the old way, we may need to fix this or change the way we count squashed reads.

Reply:
I remember looking at it and thinking it didn't change the behavior -- sorry for misleading you. We should go back to the local variable, then.